### ISSUE TITLE
some notice (Awesome project)

### ISSUE THREAD

**Issue by kotee4ko** [Martyx00/VulnFanatic#5]
Hey Bro. Thank you for you're job! It's really help me to up my skills in binary research.

if you allow i want to say few words:

1 - it will be coool if you add some searching of one-byte-overflow. 
for example when we use memcpy(var1,va2,len) and before len = strlen(some_heap_address)
something like that. this will reduce false-negative.

1.5 - i think it's possible to add some additional check to false-positive when we found something.
for example when we control var2 in memmove(var1, var2, len), but len is constant. or, len = &var1 - &var2 for example and &var1 > &var2

2 - if we try to analize some binary like "lighthttpd" - we get a fall. i dont know why :) but i can give some more log if needed.

Thank you again. Gr33t1ng5 fr0m Ukr41n3 :+1:

**Comment by Martyx00** [comment]
Hello :)
Thanks for the feedback.

Some replies:
1. I have new way of tracing in my mind and will be testing it in upcoming final (it will never be final but lest say first really usable) release of use-after-free scanner which you will notice as the version of the plugin will change to 3.0 :) This new version should be more capable of detecting issues like off-by-one (in theory).

1.5 This should already be there as for example `memmove` test case looks like this (I know the query language is stupid, but I suck as a developer so its the best thing I could have came up with 🤣 ):
```
_and_(
    _or_(
        sources.get().param(1).source_type("parameter").exported().same_branch(),
        sources.get().param(1).function_calls(["gets","scanf","fgets","recv","recvfrom", "recvmsg","getc","fread","read","sprintf","memcpy","strcpy","strncpy","getenv"]).same_branch()
    ),
    sources.get().param(2).not_constant(), <= checking that third parameter of memmove is not constant
    sources.get().param(2).not_affected_by_function_calls(["strlen","alloc"]) <= checking that third parameter is not part of alloc functions or strlen
)
```
This is not a 100% reliable for sure. But it should rule out 90% of the cases.

2. Can you share the exact binary with me please? :)

Thanks :)

**Comment by Martyx00** [comment]
I have checked with ADM64 version of `lighthttpd` from Debian official package and all went fine. Are you using latest version (v2.1)?

**Comment by kotee4ko** [comment]
![image](https://user-images.githubusercontent.com/62266948/85955216-b2379f00-b985-11ea-877e-27c12779c092.png)

since github does not support attaching binary files - i uploaded it to sendspace.
https://www.sendspace.com/file/c8m92o

i build it with make and -g key to add source code to further debug purposes.
lighttpd-1.4.55

**Comment by Martyx00** [comment]
Thanks. I will be off for one week now. But I will have a look when I am back 🙂

**Comment by Martyx00** [comment]
I have tried to replicate the issue you are facing but have not encountered it. Can you please try with the latest development release of BinaryNinja? To me it seems like bug in Binja, which was probably resolved somewhere around 2.0.2238-dev.

Let me know the result after :)

Edit: Just confirmed on the latest release of Binary Ninja dev branch (2.0.2257-dev) that this binary is scanned without any issues. Let me know the version of the Binary Ninja, if update to latest dev branch will solve the issue for you I will close this issue.

**Comment by Martyx00** [comment]
A new release have been created which requires Binary Ninja version 2263 where this issue no longer occurs (as far as my testing goes). Several improvements to tracing were also implemented and are published in this release. Closing the issue :)
